Adds global `from` support

gpopoteur opened the pull request:
This PR is based on the branch on PR #41 that bumps the Illuminate package version.

Laravel has a global from feature when sending emails, this is a feature that lets you specified the name and email address that will be used for the `from` field in all emails that the application sends when no `from` gets specified in the message.,

This PR adds support for that global from functionality.

mattstauffer commented:
I like the idea of MailThief honoring Laravel's global `from`, so that people can assert that certain routes either are or aren't overriding that.

However, explicitly/manually setting `globalFrom()` misses the purpose of that, because then you're primarily testing that you remembered to set that.

To me, this feature would be a lot more valuable if, instead of us having to explicitly set global `from`, we could just get MailThief to read the global `from` state from Laravel's config. Do you think that's something that's A) possible and B) something you're interested in doing? Thanks!

lukasoppermann commented:
@gpopoteur any news on this? It would definitely help me, as I am having the same problem.

gpopoteur commented:
@mattstauffer this PR does exactly that, on the `MailThief::hijack()` method it loads the value of the global `from` from Laravel's config file ( https://github.com/tightenco/mailthief/pull/42/files#diff-fa20f6279828b02402d6eec248b44e6aR35 ) and then after it reads the value, it calls the `$this->alwaysFrom()` method if needed.

The idea of manually calling the `$this->alwaysFrom()` method on the test is to write tests that are decoupled from the Laravel framework, at the end of the day the only method that should be coupled to Laravel is the `MailThief::hijack()` method :)

---
@lukasoppermann: this is my response to the issue, hopefully it gets merged soon as I normally use the global `from` when sending emails :)

mattstauffer commented on src/MailThief.php:
```diff
@@ -31,20 +31,26 @@ public function hijack()
         Mail::swap($this);
         app()->instance(Mailer::class, $this);
+
+        $from = app()['config']['mail.from'];
```
Would it be possible to use `array_get`?

Something like:

```php
if (config('mail.from.address')) {
    $this->alwaysFrom(config('mail.from.address'), config('mail.from.name'));
}
```

mattstauffer commented on src/MailThief.php:
```diff
@@ -31,20 +31,26 @@ public function hijack()
     {
         $message = Message::fromRaw($text);
-        $callback($message);
+        $message = $this->prepareMessage($message, $callback);
```
I don't *think* we need to assign this back to the message because of how passing objects works. Can you try it out just dropping the `$message` assignment?

mattstauffer commented:
@gpopoteur Thanks for the clarification! I didn't see a test regarding global configuration so I figured you hadn't honored it. I've added a few notes here, and can you also add a test that shows that it's reading from the config correctly? Thanks!

gpopoteur commented:
@mattstauffer there is no point in bringing in the whole Laravel Foundation package to test if the `config()` function is reading from the config file, I believe testing this is a responsibility of the framework. 

What I did here was to create a `helpers.php` file with a mocked version of the `config()` helper, I moved the logic that reads from the `config()` function into a method called `->loadGlobalFrom()`, this method is then called inside the `MailThief::hijack()` method. The idea here is to make sure that `->loadGlobalFrom()` loads the data from the `config()` function.

mattstauffer commented on src/MailThief.php:
```diff
@@ -151,4 +149,11 @@ public function alwaysFrom($address, $name = null)
         $this->from = ['address' => $address, 'name' => $name];
     }
+
+    public function loadGlobalFrom()
```
I'm not a big fan of making something a public method just for the sake of testing. It's not really a part of the public API, right?

mattstauffer commented:
@gpopoteur Sorry, I should've been more clear: you're right that we don't need to bring in the whole framework. Normally in this context I would inject a Config Manager, and in my test i could mock it.

I'm curious about your `helpers.php` approach. Let me think a bit.

Thank you so much for your hard work on this!

gpopoteur commented:
@mattstauffer I totally agree that `loadGlobalFrom()` should not be public, just made a commit changing it.

mattstauffer commented:
@gpopoteur Unfortunately changing it to protected without changing how you're calling it breaks the test.

Here's how I think I feel comfortable with it. Please let me know if you're up for this; if not, I'll take a whack at it.

I'd update the `config()` references to instead point to an injected configuration manager, and then mock that in the test.

Let me know if this is something you're interested in doing. Thanks!

gpopoteur commented:
@mattstauffer that is something I can work on, will be making the change.

gpopoteur commented:
@mattstauffer here are the latest changes I added to the PR:

- Separated the MailThief::hijack() logic into two methods (swapMail, loadFromGlobal) for better testability. Now the method that access the Laravel app can be mocked.
- Adds common tests methods used for preparing mocks into a `TestCase` abstract class.
- Config Repository is now injected into the MailThief class constructor.
- helpers.php file is now deleted as we don't need it anymore.
- Added tests to make sure the global `from` values are retrieved from the config file.

😀

besologic commented:
Nice work @gpopoteur! Thanks for all your efforts.

Resolves #50

gpopoteur commented:
Thanks for the merge @besologic ! 👍 😁
